### ISSUE TITLE
Proposal: alter no-children-prop to allow functions

### ISSUE THREAD

**Issue by erfanio** [yannickcr/eslint-plugin-react#1803]
If you use react router, is some situations you need to [pass a function as children](https://reacttraining.com/react-router/web/api/Route/children-func).
It would be nice to be able to use the same syntax as them.

```jsx
<Route path={to} children={({ match }) => (
  <li className={match ? 'active' : ''}>
    <Link to={to} {...rest}/>
  </li>
)} />
```
is easier to reason about than
```jsx
<Route path={to}>
  {({ match }) => (
    <li className={match ? 'active' : ''}>
      <Link to={to} {...rest}/>
    </li>
  )}
</Route>
```

There could be a config to allow functions (not the default).

**Comment by ljharb** [comment]
If that option were added, it’d have to disallow functions in the child position as well.

**Comment by erfanio** [comment]
Agreed. 99% of the time you'd want a pure component instead of a function anyway, so disallowing it would make sense.
